### DISCUSSION TITLE
Intercept SELECT statement to add WHERE clauses

### DISCUSSION THREAD

**Discussion by cachapa** [simolus3/drift#2560] [Q&A] [answered]
I'm trying to intercept a SELECT statement coming from the user, and add a couple WHERE clauses to transparently reduce the scope of the query without affecting its intended use.

So a simple query:
```sql
SELECT * FROM mytable
```
would become
```sql
SELECT * FROM mytable WHERE modified > ?1 AND node_id = ?2
```
---
Queries already containing where clauses should be supported too:
```sql
SELECT * FROM mytable WHERE user_id = ?1
```
would become
```sql
SELECT * FROM mytable WHERE modified > ?2 AND node_id = ?3 AND user_id = ?1
```

---

My approach is to take the parsed user sql and add my clauses to `SelectStatement.where` but haven't been able to create `BinaryExpression.operator` tokens because the `Token` constructor expects a `FileSpan` that I don't know how to instantiate.

I'm also not sure if it's good practice to modify the parsed statement, or if I should create a new copy using the `SelectStatement` constructor.

Anyway, here's my test code:

```dart
final statement = SqlEngine()
  .parse("SELECT * FROM test")
  .rootNode as SelectStatement;

statement.where = BinaryExpression(
  Reference(columnName: 'node_id'),
  Token(TokenType.equal, ???),    // Problem
  StringLiteral('12345'),
);

print(statement.toSql());
```

**Comment by simolus3** [comment, marked as answer]
The approach of mutating the statement to change the where clause sounds  good - you don't need to copy the AST.

For the tokens, I'm using this trick to create a span for them:

https://github.com/simolus3/drift/blob/f857cb17b57fd8c8aecf09d4328335cc78f3e4b2/sqlparser/test/parser/utils.dart#L9-L13

I'm happy to take contributions reducing the amount of required token references in the AST, but since this hack works I've never seen a necessity to fix it myself.

**Comment by cachapa** [comment]
Thanks, that worked!

I'd be happy to contribute a cleaner solution but am pressed for time myself, so no promises.
